src/esp/gfx/PbrShader.cpp

```diff
@@ -0,0 +1,484 @@
+}
+
+PbrShader& PbrShader::bindMetallicTexture(Mn::GL::Texture2D* texture) {
+  CORRADE_ASSERT(flags_ & Flag::MetallicTexture,
+                 "PbrShader::bindMetallicTexture: the shader was not "
+                 "created with metallic texture enabled",
+                 *this);
+  if (lightCount_ && texture) {
+    texture->bind(TextureUnit::Metallic);
+  }
+  return *this;
+}
+
+PbrShader& PbrShader::bindNormalTexture(Mn::GL::Texture2D* texture) {
+  CORRADE_ASSERT(flags_ & Flag::NormalTexture,
+                 "PbrShader::bindNormalTexture: the shader was not "
+                 "created with normal texture enabled",
+                 *this);
+  if (lightCount_ && texture) {
+    texture->bind(TextureUnit::Normal);
+  }
+  return *this;
+}
+
+PbrShader& PbrShader::bindCombinedTexture(Magnum::GL::Texture2D* texture) {
+  CORRADE_ASSERT((flags_ & Flag::OcclusionRoughnessMetallicTexture) ||
+                     (flags_ & Flag::NoneRoughnessMetallicTexture),
+                 "PbrShader::bindCombinedTexture: the shader was not "
+                 "created with combined texture enabled",
+                 *this);
+  if (lightCount_ && texture) {
+    texture->bind(TextureUnit::Combined);
+  }
+  return *this;
+}
+
+PbrShader& PbrShader::setMVPMatrix(const Mn::Matrix4& matrix) {
+  setUniform(mvpMatrixUniform_, matrix);
+  return *this;
+}
+
+PbrShader& PbrShader::setNormalMatrix(const Mn::Matrix3& matrix) {
+  setUniform(normalMatrixUniform_, matrix);
+  return *this;
+}
+
+PbrShader& PbrShader::setTransformationMatrix(const Mn::Matrix4& matrix) {
+  setUniform(modelviewMatrixUniform_, matrix);
+  return *this;
+}
+
+PbrShader& PbrShader::setObjectId(unsigned int objectId) {
+  if (flags_ & Flag::ObjectId) {
+    setUniform(objectIdUniform_, objectId);
+  }
+  return *this;
+}
+
+PbrShader& PbrShader::setBaseColor(const Mn::Color4& color) {
+  if (lightCount_) {
+    setUniform(baseColorUniform_, color);
+  }
+  return *this;
+}
+
+PbrShader& PbrShader::setRoughness(float roughness) {
+  if (lightCount_) {
+    setUniform(roughnessUniform_, roughness);
+  }
+  return *this;
+}
+
+PbrShader& PbrShader::setMetallic(float metallic) {
+  if (lightCount_) {
+    setUniform(metallicUniform_, metallic);
+  }
+  return *this;
+}
+
+PbrShader& PbrShader::bindTextures(
+    Mn::GL::Texture2D* baseColor,
+    Mn::GL::Texture2D* roughness,
+    Mn::GL::Texture2D* metallic,
+    Mn::GL::Texture2D* noneRoughnessMetallic,
+    Mn::GL::Texture2D* occlusionRoughnessMetallic,
+    Mn::GL::Texture2D* normal) {
+  (*this)
+      .bindBaseColorTexture(baseColor)
+      .bindRoughnessTexture(roughness)
+      .bindMetallicTexture(metallic)
+      .bindCombinedTexture(noneRoughnessMetallic)
+      // NO Worries to call to "bind combined texture" twice,
+      // ONLY ONE texture will be bound.
+      // see generateCorrectFlags() for details
+      .bindCombinedTexture(occlusionRoughnessMetallic)
+      .bindNormalTexture(normal);
+
+  return *this;
+}
```

Comment:
Hmm, I assume you have modelled this API after Phong? There it is primarily to avoid multiple `glActiveTexture()` + `glBindTexture()` calls and do all binding in just a [single multi-bind call](https://doc.magnum.graphics/magnum/classMagnum_1_1GL_1_1AbstractTexture.html#a9daa1b4a53fa02937aed4eaa3720dbc5) as all those binding points are consecutive. But here you're calling the single-bind APIs and thus don't really gain anything.

To be frank I don't really like that API in Phong because the texture order is unclear and error-prone, especially when a new texture type gets added. To eventually fix that there I was thinking about making the `TextureUnit` enum a public API and then having something like

```cpp
shader.bindTextures({
    {PbrShader::TextureUnit::BaseColor, baseColor},
    {PbrShader::TextureUnit::Roughness, roughness},
    {PbrShader::TextureUnit::Metallic, metallic}
    ...
});
```

That's verbose, but clear and not prone to errors. Internally the function would put those in a fixed-size array in correct order and pass that to the list-taking `bind()`.

Reply:
Good to know. Thanks. 
Yes, it did not gain anything in terms of performance. just a very convenient function to use, and make the code looks cleaner.

I will learn from the doc you mentioned above. See what I can do.

Reply:
Apart from the perf aspect, I'm just afraid it becomes rather messy once you have also occlusion and emissive textures (and then possibly other packing variants such as two-component normals), at that point it'd be very hard to make sure that the caller got the order right.

Reply:
> at that point it'd be very hard to make sure that the caller got the order right.

That is true. It is the developer's ability to make sure the order is right.

Adding comments may help in general, but not needed here, as the variable names are already self-explanatory. 
```
foo(texture1,  // A
    texture2);  // B
```

Reply:
Yes, I mean the case where someone would change that to

```cpp
foo(textureB, // B
    textureA); // A
```

because of refactoring or some other sweeping change. It still *looks* correct, but you have no way to know until you see the actual signature.

Reply:
Yes, you are right. What is you suggestion here?
How should I change the signature at this moment?

Reply:
I was thinking this in the comment above:

> ```c++
> shader.bindTextures({
>     {PbrShader::TextureUnit::BaseColor, baseColor},
>     {PbrShader::TextureUnit::Roughness, roughness},
>     {PbrShader::TextureUnit::Metallic, metallic}
>     ...
> });
> ```

Or just ... for the first iteration keep only the single-texture functions, we can always add optimizations later ;)

Reply:
I have not worked on this yet. Will do later.